src/main/java/org/springframework/data/redis/listener/RedisMessageListenerContainer.java

```diff
@@ -689,36 +690,55 @@ private class SubscriptionTask implements SchedulingAwareRunnable {
 		private class PatternSubscriptionTask implements SchedulingAwareRunnable {
 
 			private long WAIT = 500;
-			private long ROUNDS = 3;
+			private AtomicBoolean isThreadRunning = new AtomicBoolean(true);
+
 
 			public boolean isLongLived() {
 				return false;
 			}
 
+			void cancel() {
+				isThreadRunning.set(false);
+			}
+
 			public void run() {
 				// wait for subscription to be initialized
 				boolean done = false;
 				// wait 3 rounds for subscription to be initialized
-				for (int i = 0; i < ROUNDS && !done; i++) {
+				while(!done && isThreadRunning.get() && !Thread.currentThread().isInterrupted()) {
 					if (connection != null) {
-						synchronized (localMonitor) {
-							if (connection.isSubscribed()) {
-								done = true;
-								connection.getSubscription().pSubscribe(unwrap(patternMapping.keySet()));
-							} else {
-								try {
-									Thread.sleep(WAIT);
-								} catch (InterruptedException ex) {
-									Thread.currentThread().interrupt();
+							try {
+								if (connection.isSubscribed()) {
+									synchronized (localMonitor) {
+										connection.getSubscription().pSubscribe(unwrap(patternMapping.keySet()));
+										done = true;
+									}
+								} else {
+									try {
+										Thread.sleep(WAIT);
+									} catch (InterruptedException ex) {
+										logger.info("PatternSubscriptionTask was interrupted, exiting.");
+										Thread.currentThread().interrupt();
+										return;
+									}
+								}
+							} catch(Throwable e) {
+								if (e instanceof RedisConnectionFailureException) {
```

Comment:
connection errors will now be retried, like it is on the `SubscriptionTask`. This could sneakily exit before and without proper exception handling, the stack would be written to the application `stderr` and could go unnoticed.